### ISSUE TITLE
forecasting using LightGBM

### ISSUE THREAD

**Issue by nyk2001** [sktime/sktime#3784]
### Discussed in https://github.com/sktime/sktime/discussions/3783

<div type='discussions-op-text'>

<sup>Originally posted by **nyk2001** November 14, 2022</sup>
I have used ARIMA and Prophet algorithms for multivariate forecasting, which works fine.  You can see the plot with Prophet as follows:

![image](https://user-images.githubusercontent.com/3079053/201594068-3b15afa1-ddc5-4e1c-b866-d2deab379f45.png)

Now, I am trying to use LightGBM to make predictions and it does not work. For instance, I am getting the following output:

![image](https://user-images.githubusercontent.com/3079053/201594207-89d4df80-c6e7-4b31-a66d-c490b3ac42e7.png)

Is it a bug ? I am only getting problem with LightGBM and GradientBooster regressor. My code is as follows: 

`
# Load data
df = sm.datasets.macrodata.load_pandas()['data']
df['q_date'] = df.apply(lambda x: str(int(x['year'])) + '-Q' + str(int(x['quarter'])), axis=1)
df['date'] = pd.PeriodIndex(df['q_date'], freq='Q').to_timestamp()
df.loc[:, 'realinv_lagged'] = df.loc[:, 'realinv'].shift()
df[['realinv_lagged']] = df[['realinv_lagged']].fillna(method='backfill')

# Split into x and y
y = df['realgdp']
X = df[['realinv_lagged','cpi','m1']] #could be more features

# Split dataset
y_train, y_test, X_train, X_test = temporal_train_test_split(y, X, test_size=30)

# Create LGBM regressor
regressor = lgb.LGBMRegressor()
forecaster = make_reduction(regressor, window_length=10, strategy="recursive", scitype="tabular-regressor")
    
# Define horizon
fh_abs = ForecastingHorizon(y_test.index, is_relative=False) 

# Fit model
forecaster.fit(y_train, X_train)

# Make predictions
y_pred = forecaster.predict(X=X_test, fh=fh_abs)

# Plot preditcions
plot_series(y_train, y_test, y_pred, labels=["y_train", "y_test", "y_pred"])
`
</div>

**Comment by nyk2001** [comment]
I will have a look at the article. One more question !! 

In your documentation examples, you have used GradientBooster for forecasting. I have used the same code but it also does not work. Why is that ?

https://www.sktime.org/en/stable/api_reference/auto_generated/sktime.forecasting.compose.make_reduction.html?highlight=make_reduction

**Comment by danbartl** [comment]
As i said, by design.gradient boosting trees cannot extrapolate by design. They will NOT forecast values they have not seen. So the airlines dataset is an extremely bad example since it has a trend. You need to do decomposition to get useful forecasts, read the example.

**Comment by nyk2001** [comment]
Thanks for the clarification

**Comment by nyk2001** [comment]
> As i said, by design, gradient boosting trees cannot extrapolate . They will NOT forecast values they have not seen. So the airlines dataset is an extremely bad example since it has a trend. You need to do decomposition to get useful forecasts, read the article.

One more thing. The sample article is great for LightGBM. Interestingly, I am planning to use the same dataset. In his example, he limited the use to just 1 store because if you include all stores then there will be date duplicates. That is going to impact the forecast horizon. Because, I am going to use Test Dataset index (date) as my forecast horizon for testing purposes. 

Is there a way to handle this in SkTime ? Such as combine date with store and item id and use it as dataframe index ?

**Comment by aiwalter** [comment]
https://github.com/sktime/sktime-tutorial-pydata-berlin-2022/blob/main/notebooks/3_hierarchical_global_forecasting.ipynb

**Comment by nyk2001** [comment]
@aiwalter thanks for sharing the link.  In note book examples, it is assumed that you have data in the desired format. 

I am struggling to find examples where I can convert a 2D pandas dataframe into hierarchical sktime compatible format. Can you please share some example ? 

I want to use store, item, and timestamp as indices. My pandas DatafFame looks like as follows: 

![image](https://user-images.githubusercontent.com/3079053/201812877-2e92e839-cd5f-4dcc-80bc-969b72480551.png)

Any guidance will be really appreciated. I am new to timeseries domain !!

**Comment by fkiraly** [comment]
@nyk2001, maybe one additional thing: yes, ensembles of trees do not extrapolate well, but one simple trick to fix that is applying a transformation to make the time series more stationary and back-transform.

In your example, it would be a pipeline with the differencer, e.g.,

```python
forecaster = make_reduction(regressor, window_length=10, strategy="recursive", scitype="tabular-regressor")
forecaster_with_differencer = Differencer() * forecaster
```

**Comment by nyk2001** [comment]
@fkiraly Thank you. At the moment, I am trying to find a way to convert my dataframe to above format. Please let me know if you have experimented it.

**Comment by nyk2001** [comment]
@fkiraly @danbartl Thanks for your inputs. I can successfully use LigthGBM model to do the forecasting.
